text/0000-prelude-2021.md

```diff
@@ -0,0 +1,164 @@
+- Feature Name: `prelude_2021`
+- Start Date: 2021-02-16
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+A new prelude for the 2021 edition, featuring several extra traits.
+
+# Motivation
+[motivation]: #motivation
+
+While types and free functions can be added to the prelude independent of edition boundaries, the same is not true for traits. Adding a trait to the prelude can cause compatibility issues because calls to methods named the same as methods of the newly in scope traits can become ambiguous. Because editions are opt-in, they represent an opportunity to add new traits to the prelude.
+
+The expected outcome is to make the newly added traits more accessible to users, potentially stimulating their use compared to alternatives.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The compiler currently brings all items from `std::prelude::v1` into scope for each module (for code using `std`). Crates that declare usage of the 2021 edition will instead get items from `std::prelude::edition2021` imported into scope.
+
+This RFC proposes to add the following traits to `std::prelude::edition2021`, but not `std::prelude::v1`:
+
+- `TryFrom`/`TryInto`
+- `FromIterator`
+
+Apart from the 2021-specific prelude, this RFC also proposes to add several items to preludes for all editions. These changes do not have to be aligned with an edition boundary, since there is no compatibility hazard (explicitly imported names always shadow prelude names).
+
+- `std::{mem, fmt}`
+- `std::collections::{HashMap, HashSet}`
+- `std::path::{Path, PathBuf}`
+- `std::sync::{Arc, Mutex, RwLock}`
+- `std::borrow::Cow`
+- `std::fs::File`
```

Comment:
`File` doesn't seem very useful without `Read` or `Write`. It also has the same problem with async as `Mutex` and `RwLock`.

Reply:
Though on that note, having `Read`/`Write` in the prelude would make a **lot** of APIs a lot easier to use. Socket and File IO are frequently annoying to use due to having to import both the traits and the concrete type (File/TcpStream).

Reply:
I would second adding `Read`/`Write` to the `std::prelude`, yes

Reply:
Maybe Read/Write but actually not the types? Unsure considering the AsyncRead concern, but

Reply:
If `std::io::Write` is added to the prelude, using `std::fmt::Write` becomes more difficult because of name conflicts. Not sure how many people use that trait, tho.

Reply:
They could be added as `use std::io::{Read as _, Write as _};`

Reply:
If the prelude includes `std::io::Write`, how do I `write!` using `std::fmt::Write` to a `String`?